File: leetcode/algorithms/234_palindrome_linked_list/main.js
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * Complexities:
 *   N - The number of nodes in `head`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(1)
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function (head) {
  const obverse = [];
  const reverse = [];
  let newHead = head;

  while (newHead) {
    obverse.push(newHead.val);
    newHead = newHead.next;
  }

  let prev = null;
  let current = head;

  while (current) {
    const next = current.next;
    current.next = prev;
    prev = current;
    current = next;
  }

  while (prev) {
    reverse.push(prev.val);
    prev = prev.next;
  }

  for (let i = 0; i < obverse.length; i++) {
    if (obverse[i] !== reverse[i]) {
      return false;
    }
  }

  return true;
};


// Solution
/**
 * Complexities:
 *   N - The number of nodes in `head`
 *   - Time Complexity: O(N)
 *   - Space Complexity: O(1)
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var solution = function (head) {
  if (!head || !head.next) {
    return true;
  }

  let slow = head;
  let fast = head;

  while (fast !== null && fast.next !== null) {
    slow = slow.next;
    fast = fast.next.next;
  }

  let prev = null;
  let curr = slow;

  while (curr !== null) {
    const nextTemp = curr.next;
    curr.next = prev;
    prev = curr;
    curr = nextTemp;
  }

  let p1 = head;
  let p2 = prev;

  while (p2 !== null) {
    if (p1.val !== p2.val) {
      return false;
    }
    p1 = p1.next;
    p2 = p2.next;
  }

  return true;
};

File: leetcode/algorithms/234_palindrome_linked_list/main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

class PalindromeLinkedList {
    class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    /**
     * Complexities:
     *   N - The number of nodes in `head`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    public boolean isPalindrome(ListNode head) {
        ListNode newHead = head;
        List<Integer> obverse = new ArrayList<>();
        List<Integer> reverse = new ArrayList<>();

        while (newHead != null) {
            obverse.add(newHead.val);
            newHead = newHead.next;
        }

        ListNode prev = null;
        ListNode current = head;

        while (current != null) {
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        while (prev != null) {
            reverse.add(prev.val);
            prev = prev.next;
        }

        for (int i = 0; i < obverse.size(); i++) {
            if (!Objects.equals(obverse.get(i), reverse.get(i))) {
                return false;
            }
        }

        return true;
    }


    // Solution
    /**
     * Complexities:
     *   N - The number of nodes in `head`
     *   - Time Complexity: O(N)
     *   - Space Complexity: O(1)
     */
    public boolean solution(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }

        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode secondHalfHead = reverse(slow);

        ListNode p1 = head;
        ListNode p2 = secondHalfHead;
        boolean isPalindrome = true;

        while (p2 != null) {
            if (p1.val != p2.val) {
                isPalindrome = false;
                break;
            }
            p1 = p1.next;
            p2 = p2.next;
        }

        reverse(secondHalfHead);

        return isPalindrome;
    }

    private ListNode reverse(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;

        while (curr != null) {
            ListNode nextTemp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextTemp;
        }

        return prev;
    }
}
